File: src/main/variations/appearances/Animation.java
package main.variations.appearances;

import main.variations.Variation;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.ImmutablePair;
import processing.core.PImage;
import processing.data.JSONArray;
import processing.data.JSONObject;

import java.util.ArrayList;

public class Animation extends Appearance {

    private ArrayList<PImage> images = new ArrayList<>();
    private int time; //time between frames
    private int lastTime;
    private int index;
    private ArrayList<String> paths = new ArrayList<>();
    private ArrayList<String> names = new ArrayList<>();

    public Animation(Variation variation, JSONObject json) {
        super(variation);
        JSONArray textures = json.getJSONArray("textures");
        for (int i = 0;i < textures.size();i++) {
            String path = textures.getString(i);
            if (StringUtils.countMatches(path, ":") < 2) {
                path = this.getVariation().getElement().getName() + ":" + path;
            }
            this.paths.add(path);
            this.images.add(this.getVariation().getElement().getImage(path));
        }
        JSONArray names = json.getJSONArray("names");
        if (names != null) {
            for (int i = 0; i < textures.size(); i++) {
                if (i < names.size()) {
                    this.names.add(names.getString(i));
                }
            }
        }
        this.time = json.hasKey("time") ? json.getInt("time") : 1000;
    }

    @Override
    public String getName() {
        return this.index < this.names.size() ? this.names.get(this.index) : null;
    }

    @Override
    public PImage getImage() {
        if (this.lastTime + this.time <= main.millis()) {
            this.lastTime = main.millis();
            this.index++;
            if (this.index >= this.images.size()) {
                this.index = 0;
            }
        }
        return this.images.get(this.index);
    }

    @Override
    public ArrayList<ImmutablePair<PImage, String>> getPairs() {
        ArrayList<ImmutablePair<PImage, String>> list = new ArrayList<>();
        for (int i = 0;i < this.images.size();i++) {
            if (this.images.get(i) != null) {
                list.add(new ImmutablePair<>(this.images.get(i), this.paths.get(i)));
            }
        }
        return list;
    }
}

File: src/main/variations/appearances/Animation.kt
package main.variations.appearances

import main.variations.Variation
import org.apache.commons.lang3.StringUtils
import processing.core.PImage
import processing.data.JSONObject
import java.util.*

class Animation(val variation: Variation, json: JSONObject) : Appearance() {
    private val images = ArrayList<PImage>()
    private val time //time between frames
            : Int
    private var lastTime = 0
    private var index = 0
    private val paths = ArrayList<String>()
    private val names = ArrayList<String>()
    override fun getName(): String = if (index < names.size) names[index] else "null"
    override fun getImage(): PImage {
        if (lastTime + time <= main.millis()) {
            lastTime = main.millis()
            index++
            if (index >= images.size) {
                index = 0
            }
        }
        return images[index]
    }

    override fun getPairs(): List<Pair<PImage, String>> {
        //TODO what is the path for strip animation textures?
        return images.zip(paths)
    }

    init {
        if (json.hasKey("texture")) {
            val image = variation.element.loadImage(json.getString("texture"))
            for (i in 0 until image.height / image.width) {
                images.add(image.get(0, i * image.width, image.width, image.width))
            }
        } else {
            val textures = json.getJSONArray("textures")
            for (i in 0 until textures.size()) {
                var path = textures.getString(i)
                if (StringUtils.countMatches(path, ":") < 2) {
                    path = "${this.variation.element.id}:$path"
                }
                paths.add(path)
                images.add(this.variation.element.loadImage(path))
            }
            json.getJSONArray("names")?.let {
                for (i in 0 until textures.size()) {
                    if (i < it.size()) {
                        this.names.add(it.getString(i))
                    }
                }
            }
        }
        time = json.getInt("time", 1000)
    }
}